api/renter.go

```diff
@@ -350,49 +379,131 @@ func (api *API) renterDeleteHandler(w http.ResponseWriter, req *http.Request, ps
 func (api *API) renterDownloadAsyncHandler(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
-	destination := req.FormValue("destination")
-	// Check that the destination path is absolute.
-	if !filepath.IsAbs(destination) {
-		WriteError(w, Error{"destination must be an absolute path"}, http.StatusBadRequest)
+	// Ensure that the async flag is not set to any other value than "true" or "".
+	async := req.FormValue("async")
+	if async == "false" || async != "true" && async != "" {
+		WriteError(w, Error{"/downloadasync require the `async` parameter to be either equal to `true` or be empty."}, http.StatusBadRequest)
 		return
+	} else if async == "" { // Add async if it is not already set to `true`.
+		req.Form["async"] = []string{"true"}
 	}
 
-	siapath := strings.TrimPrefix(ps.ByName("siapath"), "/")
-	fileExists := false
+	api.renterDownloadHandler(w, req, ps)
+}
 
-	files := api.renter.FileList()
-	for _, file := range files {
-		if file.SiaPath == siapath {
-			fileExists = true
+func (api *API) parseDownloadParameters(w http.ResponseWriter, req *http.Request, ps httprouter.Params) (*modules.RenterDownloadParameters, error) {
+	destination := req.FormValue("destination")
+
+	// The offset and length in bytes.
+	offsetparam := req.FormValue("offset")
+	lengthparam := req.FormValue("length")
+
+	// Determines whether the response is written to response body.
+	httprespparam := req.FormValue("httpresp")
+
+	// Determines whether to return on completion of download or straight away.
+	// If httprespparam is present, this parameter is ignored.
+	asyncparam := req.FormValue("async")
+
+	// Parse the offset and length parameters.
+	var offset, length uint64
+	var err error
+	if len(offsetparam) > 0 {
+		offset, err = strconv.ParseUint(offsetparam, 10, 64)
+		if err != nil {
+			return nil, build.ExtendErr("could not decode the offset as uint64: ", err)
 		}
 	}
+	if len(lengthparam) > 0 {
+		length, err = strconv.ParseUint(lengthparam, 10, 64)
+		if err != nil {
+			return nil, build.ExtendErr("could not decode the length as uint64: ", err)
+		}
+	}
+	// Verify that if either offset or length have been provided that both were provided.
+	offparampassed := len(offsetparam) > 0
+	lenparampassed := len(lengthparam) > 0
 
-	if !fileExists {
-		WriteError(w, Error{"file at specified siapath does not exist"}, http.StatusBadRequest)
-		return
+	// Parse the httpresp parameter.
+	httpresp, err := stringToBool(httprespparam)
+	if err != nil {
+		return nil, build.ExtendErr("httpresp parameter could not be parsed", err)
+	}
+
+	// Parse the async parameter.
+	async, err := stringToBool(asyncparam)
+	if err != nil {
+		return nil, build.ExtendErr("async parameter could not be parsed", err)
+	}
+
+	// Verify that either async or httpresp are true but not both.
+	if async && httpresp {
+		return nil, errors.New("only one of the async and httpresp flags can be specified.")
 	}
```

Comment:
Just FYI, I refactored this code to move all checks into the download function in download.go that can be moved there. These lines could be moved into the download handler functions since that is where they are used.